### ISSUE TITLE
Add a __repr__ for Figure without the object address ?

### ISSUE THREAD

**Issue by Carreau** [matplotlib/matplotlib#9372]
This is not really a bug report ; but mostly to start a discussion. 

@fperez is teaching reproducible research at UC Berkeley and we are attempting to get a byte for byte reproducibility of notebooks. One of the object repr that change the most often are the Figure REPRs because of the memory address in the REPR. 

The tricky part is as all the mimetypes representations are stored in the notebook, even if the hex memory location is not shown to the user, it is stored in the notebook. For for every figure displayed inline there is a text representation that change after every run. 

As `Figure` already have a `__str__`,  would it make sens to also add a `__repr__` ? And at the same time if accepted; make it more informative like the number of axes it contains for example ?

**Comment by anntzer** [comment]
How do you deal with, say, `object()`, which also includes the address?

**Comment by Carreau** [comment]
I don't deal with it. The issue is that the `__repr__` is shown for every figures whn using `%matpltolib inline`. Most other objects don't ave an hex repr. 

The issue with Figures is that as they have both the `__repr__` and the `_repr_png_,` visually for the user it is reproducible (PNG is identical), but as the notebook also store the text repr the notebook is not identical.

**Comment by anntzer** [comment]
Does jupyter really need to store that repr (not very familiar with the internals)?

**Comment by Carreau** [comment]
> Does jupyter really need to store that repr (not very familiar with the internals)?

It does store it, we don't really just change the spec as many tools downstream do expect that, and that would be special casing `plain/text` mimetype only for an edge case. We can also overwrite the Matplotlib repr in IPython if necessary, but I just wanted to bring also the discussion on the matplotlib side before doing anything.

**Comment by anntzer** [comment]
I don't have a strong stance against changing it, but I still don't understand how this is a mpl-specific issue.

**Comment by Carreau** [comment]
> I don't have a strong stance against changing it, but I still don't understand how this is a mpl-specific issue.

It's not completly MPL **specific**, it affect any object which main use is not textual and keep the default `__repr__`, so I'm pretty sure I can find other library doing this.

 If I can make a crude comparison is like zip files that embed the timestamp of creation. The actual content of the zip is strictly identical, but the hash of the zip change every-time you recreate it.

Most people won't see it, but if you want reproducible build it's annoying. It's exacerbated in matplotlib figure because many notebook contain a lot of these so it is more noticeable.

**Comment by anntzer** [comment]
Honestly given the way the problem is stated (if you don't want to strip reprs yourself) I'm tempted to tag as wontfix (because you are effectively asking the entire Python ecosystem to never leak a repr containing an address into jupyter -- if it is really so, this should just be fixed at the cpython level...).  But I'm happy to let another core dev chime in.

**Comment by tacaswell** [comment]
I am :+1: on implementing this, the hex repr is basically useless (and I look at it a lot in the terminal).

Exact work:

 - add `__repr__` method to `Figure`.  The `__str__` method seems to show the size in pixels, I think figure name is probably more useful (in both?).
 - maybe change `__str__` as well.

**Comment by fperez** [comment]
+1, @tacaswell - I figured this would be an opportunity to:

- improve bit-wise reproducibility of notebook runs. This would be in line both with efforts for bit-wise reproducibility in the packaging community for example, and with MPL's new support for removing timestamps from PDF/PS/SVG (thanks for that, BTW, had a chance to discuss it today in class as the PDF issue had also come up; perfect timing :)

- provide more informative and useful reprs by default (even a bit of info, like # of axes and/or sizes, could be nice to have)

Many thanks for considering it!

**Comment by fperez** [comment]
ps - while Figure is where the highest payoff-to-work ratio lies, other objects could benefit from a similar treatment, like the common

```
In [2]: plot([1,2,3])
Out[2]: [<matplotlib.lines.Line2D at 0x11e236fd0>]
```

**Comment by matthew-brett** [comment]
I don't have a moral feel on the issue, but I would say that the matplotlib object reprs are by far the most most common ugly reprs in the Notebooks I make.  So, if y'all could see a way to changing those, it would make a lot of my tutorials look a whole lot neater, and a little bit easier to explain.

**Comment by anntzer** [comment]
Given the actual goal here, instead of special casing figure, I'd start by attaching a better repr on Artist (which includes Figure), e.g. something like
```
diff --git a/lib/matplotlib/artist.py b/lib/matplotlib/artist.py
index 21edc744c..6c8214f7d 100644
--- a/lib/matplotlib/artist.py
+++ b/lib/matplotlib/artist.py
@@ -128,6 +128,15 @@ class Artist(object):
         d['stale_callback'] = None
         return d
 
+    def __repr__(self):
+        cls = type(self)
+        label = self.get_label()
+        if label:
+            return "<{}.{}: label={!r}>".format(
+                cls.__module__, cls.__name__, label)
+        else:
+            return "<{}.{}>".format(cls.__module__, cls.__name__)
+
     def remove(self):
         """
         Remove the artist from the figure if possible.  The effect
```
(perhaps that'll encourage also people to properly use labels for legends instead of the matlab-style API... :-)) and worry about specific classes later.

(mild preference for `classname: key=value` instead of `classname(key=value)` as the latter suggests that the function call would be correct).

Just to show that I don't hold any grudges :-)

**Comment by Carreau** [comment]
> However, given @fperez's last comment, I really wonder whether the jupyter devs plan to chase around each and every project in the scipy stack to make sure they fix the reprs of all their classes

Haha, no, and having people to push back is alway a good thing to see the all picture. 
As I mentioned earlier matplotlib figure are quite frequent hence why I opened this issue with this title. We _can_ overwrite in IPython without having to monkeypatch, so that's not an issue. I think we need to distinguish 2 kinds of objects:
 -  object that so not have a helpful repr, you need to be able to distinguish multiple of them. For these the hex value is critical. 
 - Object that do have a helpful representation – for these Hex seem unnecessary and the hex value unproductive.

For matplotlib, I believe that `Figures`, (and most `Artists`) should enter the second category. Figures have already nice PNG reprs. 

I did not request changes to Artists as I believed it would have been blankly rejected as too much of  a change, though if Artist reprs were to change I would be even happier.

**Comment by dopplershift** [comment]
@anntzer While I appreciate the purity of your argument, I'm 💯 on an effective solution that kills 90% of the problem right off the top. I'm very much of solutions that solve real problems quickly (provided they don't increase maintenance burden), rather than spending weeks getting buy in from the larger ecosystem. It feels a bit like this:
![xkcd condiments](https://imgs.xkcd.com/comics/the_general_problem.png)

 I think your suggestion of changing `Artist` is useful though.

**Comment by amueller** [comment]
This has been done, right?
I saw this has been done for text, but not for colorbar and legend. Would you accept a PR for that or is someone working on that already?

**Comment by QuLogic** [comment]
It seems that the main goal here was fixed in #9379.
